span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/JaegerSpanNormalizer.java

```diff
@@ -127,23 +146,80 @@ private Callable<RawSpan> getRawSpanNormalizerCallable(
 
-      // redact PII tags, tag comparisons are case insensitive
+      // redact PII tags, tag comparisons are case insensitive (Resource tags are skipped)
+      sanitiseSpan(rawSpanBuilder);
+
+      // build raw span
+      RawSpan rawSpan = rawSpanBuilder.build();
+      if (LOG.isDebugEnabled()) {
+        logSpanConversion(jaegerSpan, rawSpan);
+      }
+
+      return rawSpan;
+    };
+  }
+
+  private void sanitiseSpan(Builder rawSpanBuilder) {
+
+    try {
       var attributeMap = rawSpanBuilder.getEvent().getAttributes().getAttributeMap();
       Set<String> tagKeys = attributeMap.keySet();
 
       AtomicReference<Boolean> containsPIIFields = new AtomicReference<>();
       containsPIIFields.set(false);
 
-      tagKeys.stream()
-          .filter(tagKey -> tagsToRedact.contains(tagKey.toUpperCase()))
-          .peek(tagKey -> containsPIIFields.set(true))
-          .forEach(
-              tagKey ->
+      try {
+        tagKeys.stream()
+            .filter(tagKey -> tagKeysToRedact.contains(tagKey.toUpperCase()))
+            .peek(tagKey -> containsPIIFields.set(true))
+            .forEach(
+                tagKey -> {
+                  spanAttributesRedactedCounters
+                      .computeIfAbsent(
+                          MatchType.KEY.toString(),
+                          k ->
+                              registerCounter(
+                                  SPAN_REDACTED_ATTRIBUTES_COUNTER,
+                                  Map.of("matchType", MatchType.KEY.toString())))
+                      .increment();
+                  logSpanRedaction(tagKey, MatchType.KEY);
                   attributeMap.put(
                       tagKey,
                       AttributeValue.newBuilder()
                           .setValue(SpanNormalizerConstants.PII_FIELD_REDACTED_VAL)
-                          .build()));
+                          .build());
+                });
+      } catch (Exception e) {
+        LOG.error("An exception occurred while sanitising spans with key match: ", e);
+      }
+
+      try {
+        for (String regexPattern : tagRegexPatternToRedact) {
+          Pattern pattern =
+              regexPatternMap.computeIfAbsent(regexPattern, k -> Pattern.compile(regexPattern));
+          for (String tagKey : tagKeys) {
+            if (pattern.matcher(attributeMap.get(tagKey).getValue()).matches()) {
+              containsPIIFields.set(true);
+              spanAttributesRedactedCounters
+                  .computeIfAbsent(
+                      MatchType.REGEX.toString(),
+                      k ->
+                          registerCounter(
+                              SPAN_REDACTED_ATTRIBUTES_COUNTER,
+                              Map.of("matchType", MatchType.REGEX.toString())))
+                  .increment();
+              logSpanRedaction(tagKey, MatchType.REGEX);
+              attributeMap.put(
+                  tagKey,
+                  AttributeValue.newBuilder()
```

Comment:
Let us create a single `AttributeValue` object and reuse it everywhere. Any operations on this map later are read operations so this should be OKAY.